File: sdk/android/src/main/kotlin/io/seeray/lens/android/SeeRayAnalytics.kt
package io.seeray.lens.android

import android.content.Context
import android.content.SharedPreferences
import java.io.Closeable
import java.net.HttpURLConnection
import java.net.URI
import java.net.URL
import java.nio.charset.StandardCharsets
import java.time.Clock
import java.time.Duration
import java.time.Instant
import java.util.ArrayDeque
import java.util.Locale
import java.util.UUID
import java.util.concurrent.Executors
import java.util.concurrent.Future
import java.util.concurrent.FutureTask
import java.util.concurrent.ScheduledExecutorService
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicBoolean

data class SeeRayAnalyticsOptions(
    val siteId: String,
    val apiOrigin: String,
    val requireConsent: Boolean = false,
    val batchSize: Int = 10,
    val flushInterval: Duration = Duration.ofSeconds(10),
)

enum class AnalyticsConsent { UNKNOWN, GRANTED, DENIED }

/**
 * Privacy-conscious Android event client for SeeRay Lens.
 *
 * Call [trackPageView] when an app screen becomes visible and [trackEvent] for meaningful
 * interactions. No screen, lifecycle, advertising ID, device model, or user identity is collected
 * automatically.
 */
class SeeRayAnalytics internal constructor(
    private val options: SeeRayAnalyticsOptions,
    private val store: AnalyticsStore,
    private val transport: AnalyticsTransport,
    private val clock: Clock = Clock.systemUTC(),
    startScheduler: Boolean = true,
) : Closeable {
    constructor(context: Context, options: SeeRayAnalyticsOptions) : this(
        options = options.validated(),
        store = AndroidAnalyticsStore(context.applicationContext),
        transport = UrlConnectionAnalyticsTransport(),
    )

    private val lock = Any()
    private val queue = ArrayDeque<AnalyticsEvent>()
    private val sending = AtomicBoolean(false)
    private val closing = AtomicBoolean(false)
    private val batchSize = options.batchSize.coerceIn(1, MAX_BATCH_SIZE)
    private val endpoint = endpoint(options.apiOrigin)
    private val scheduler: ScheduledExecutorService = Executors.newSingleThreadScheduledExecutor { task ->
        Thread(task, "seeray-analytics").apply { isDaemon = true }
    }
    private var visitorId: String? = null
    private var sessionId: String? = null
    private var sessionLastActivity: Long? = null
    private var userId: String? = null
    private var droppedEventCount = 0L

    init {
        options.validated()
        if (startScheduler) {
            val interval = options.flushInterval.toMillis().coerceIn(MIN_FLUSH_MS, MAX_FLUSH_MS)
            scheduler.scheduleWithFixedDelay({ flush() }, interval, interval, TimeUnit.MILLISECONDS)
        }
    }

    fun consentState(): AnalyticsConsent = synchronized(lock) {
        when (store.get(KEY_CONSENT)) {
            CONSENT_GRANTED -> AnalyticsConsent.GRANTED
            CONSENT_DENIED -> AnalyticsConsent.DENIED
            else -> if (options.requireConsent) AnalyticsConsent.UNKNOWN else AnalyticsConsent.GRANTED
        }
    }

    /** Records an explicit visitor choice. Denial immediately removes local IDs and queued events. */
    fun setConsent(granted: Boolean) {
        synchronized(lock) {
            if (granted) {
                store.put(KEY_CONSENT, CONSENT_GRANTED)
                ensureIdentity(clock.millis())
            } else {
                store.put(KEY_CONSENT, CONSENT_DENIED)
                clearIdentity()
                queue.clear()
                userId = null
            }
        }
    }

    fun optOut() = setConsent(granted = false)

    /**
     * Sets an opaque application-owned ID only while consent is granted. The server hashes it
     * before persistence. Clear it on logout/account switch; never pass an email or other PII.
     */
    fun setUserId(value: String?) {
        synchronized(lock) {
            userId = if (consentState() == AnalyticsConsent.GRANTED) cleanUserId(value) else null
        }
    }

    fun trackPageView(url: String, title: String? = null, referrer: String? = null) = record(
        type = "page_view",
        url = url,
        title = title,
        referrer = referrer,
    )

    fun trackScreen(screenName: String, url: String, title: String? = screenName) = record(
        type = "page_view",
        url = url,
        title = title,
        properties = mapOf("screen" to screenName),
    )

    /**
     * Records a configured goal event. For named goals, prefer [trackEvent] using the event type
     * and name configured in the SeeRay goal editor.
     */
    fun trackGoal(name: String, url: String, properties: Map<String, Any?> = emptyMap()) = record(
        type = "goal",
        url = url,
        name = name,
        properties = properties,
    )

    fun trackEvent(
        eventType: String,
        url: String,
        category: String? = null,
        action: String? = null,
        name: String? = null,
        properties: Map<String, Any?> = emptyMap(),
    ) {
        val cleanType = cleanEventType(eventType) ?: return
        record(
            type = cleanType,
            url = url,
            category = category,
            action = action,
            name = name,
            properties = properties,
        )
    }

    fun pendingEventCount(): Int = synchronized(lock) { queue.size }

    fun droppedEventCount(): Long = synchronized(lock) { droppedEventCount }

    /** Sends one bounded batch off the calling thread. Failed batches remain queued for retry. */
    fun flush(): Future<Boolean> = try {
        scheduler.submit<Boolean> { sendOneBatch() }
    } catch (_: java.util.concurrent.RejectedExecutionException) {
        FutureTask<Boolean> { false }.apply { run() }
    }

    override fun close() {
        if (!closing.compareAndSet(false, true)) return
        try {
            // Drain one final batch on the worker. Never block an Activity/main-thread shutdown.
            scheduler.execute {
                try {
                    sendOneBatch()
                } finally {
                    scheduler.shutdown()
                }
            }
        } catch (_: java.util.concurrent.RejectedExecutionException) {
            scheduler.shutdownNow()
        }
    }

    private fun record(
        type: String,
        url: String,
        title: String? = null,
        referrer: String? = null,
        category: String? = null,
        action: String? = null,
        name: String? = null,
        properties: Map<String, Any?> = emptyMap(),
    ) {
        val pageUrl = validatedPageUrl(url) ?: return
        val eventType = cleanEventType(type) ?: return
        synchronized(lock) {
            if (closing.get()) return
            if (consentState() != AnalyticsConsent.GRANTED) return
            if (queue.size >= MAX_PENDING_EVENTS) {
                droppedEventCount++
                return
            }
            val now = clock.instant()
            val ids = ensureIdentity(now.toEpochMilli())
            queue.addLast(
                AnalyticsEvent(
                    eventId = UUID.randomUUID().toString(),
                    type = eventType,
                    occurredAt = now.toString(),
                    url = pageUrl,
                    title = boundedText(title, MAX_TITLE),
                    referrer = validatedPageUrl(referrer),
                    visitorId = ids.first,
                    sessionId = ids.second,
                    userId = userId,
                    category = boundedText(category, MAX_CATEGORY),
                    action = boundedText(action, MAX_CATEGORY),
                    name = boundedText(name, MAX_NAME),
                    properties = cleanProperties(properties),
                    language = Locale.getDefault().toLanguageTag().take(MAX_LANGUAGE),
                    androidVersion = android.os.Build.VERSION.RELEASE?.take(MAX_VERSION),
                ),
            )
        }
        if (pendingEventCount() >= batchSize) flush()
    }

    private fun sendOneBatch(): Boolean {
        if (!sending.compareAndSet(false, true)) return false
        val batch = synchronized(lock) {
            if (consentState() != AnalyticsConsent.GRANTED || queue.isEmpty()) {
                emptyList()
            } else {
                buildList { repeat(minOf(batchSize, queue.size)) { add(queue.removeFirst()) } }
            }
        }
        if (batch.isEmpty()) {
            sending.set(false)
            return true
        }
        val success = try {
            transport.post(endpoint, TrackingRequest(options.siteId, clock.instant().toString(), batch).toJson())
        } catch (_: Exception) {
            false
        }
        synchronized(lock) {
            if (consentState() == AnalyticsConsent.GRANTED) {
                if (success) {
                    store.put(KEY_SESSION_LAST_ACTIVITY, clock.millis().toString())
                    sessionLastActivity = clock.millis()
                } else {
                    for (event in batch.asReversed()) {
                        if (queue.size < MAX_PENDING_EVENTS) queue.addFirst(event) else droppedEventCount++
                    }
                }
            }
        }
        sending.set(false)
        return success
    }

    private fun ensureIdentity(nowMillis: Long): Pair<String, String> {
        if (visitorId == null) visitorId = store.get(KEY_VISITOR_ID) ?: UUID.randomUUID().toString().also {
            store.put(KEY_VISITOR_ID, it)
        }
        val persistedLastActivity = sessionLastActivity ?: store.get(KEY_SESSION_LAST_ACTIVITY)?.toLongOrNull()
        val currentSession = sessionId ?: store.get(KEY_SESSION_ID)
        if (currentSession == null || persistedLastActivity == null || nowMillis - persistedLastActivity >= SESSION_TIMEOUT_MS) {
            sessionId = UUID.randomUUID().toString()
            store.put(KEY_SESSION_ID, sessionId!!)
        } else {
            sessionId = currentSession
        }
        sessionLastActivity = nowMillis
        store.put(KEY_SESSION_LAST_ACTIVITY, nowMillis.toString())
        return visitorId!! to sessionId!!
    }

    private fun clearIdentity() {
        visitorId = null
        sessionId = null
        sessionLastActivity = null
        store.remove(KEY_VISITOR_ID)
        store.remove(KEY_SESSION_ID)
        store.remove(KEY_SESSION_LAST_ACTIVITY)
    }

    private data class AnalyticsEvent(
        val eventId: String,
        val type: String,
        val occurredAt: String,
        val url: String,
        val title: String?,
        val referrer: String?,
        val visitorId: String,
        val sessionId: String,
        val userId: String?,
        val category: String?,
        val action: String?,
        val name: String?,
        val properties: Map<String, Any>,
        val language: String,
        val androidVersion: String?,
    ) {
        fun toJson(): String = jsonObject(
            linkedMapOf(
                "eventId" to eventId,
                "type" to type,
                "occurredAt" to occurredAt,
                "url" to url,
                "title" to title,
                "referrer" to referrer,
                "visitorId" to visitorId,
                "sessionId" to sessionId,
                "userId" to userId,
                "category" to category,
                "action" to action,
                "name" to name,
                "properties" to properties,
                "context" to linkedMapOf(
                    "browser" to "Other",
                    "operatingSystem" to "Android",
                    "operatingSystemVersion" to androidVersion,
                    "deviceType" to "mobile",
                    "language" to language,
                ),
            ),
        )
    }

    private data class TrackingRequest(val siteId: String, val sentAt: String, val events: List<AnalyticsEvent>) {
        fun toJson(): String = jsonObject(
            linkedMapOf(
                "schemaVersion" to 1,
                "siteId" to siteId,
                "sentAt" to sentAt,
                "events" to events.map { JsonFragment(it.toJson()) },
            ),
        )
    }

    private data class JsonFragment(val value: String)

    companion object {
        private const val KEY_CONSENT = "consent"
        private const val KEY_VISITOR_ID = "visitor_id"
        private const val KEY_SESSION_ID = "session_id"
        private const val KEY_SESSION_LAST_ACTIVITY = "session_last_activity"
        private const val CONSENT_GRANTED = "granted"
        private const val CONSENT_DENIED = "denied"
        private const val MAX_PENDING_EVENTS = 100
        private const val MAX_BATCH_SIZE = 10
        private const val MAX_EVENT_TYPE = 64
        private const val MAX_TITLE = 256
        private const val MAX_CATEGORY = 80
        private const val MAX_NAME = 128
        private const val MAX_LANGUAGE = 35
        private const val MAX_VERSION = 24
        private const val SESSION_TIMEOUT_MS = 30 * 60 * 1000L
        private const val MIN_FLUSH_MS = 1_000L
        private const val MAX_FLUSH_MS = 60_000L

        private fun endpoint(apiOrigin: String): String {
            val origin = URI(apiOrigin.trim())
            require(origin.scheme.equals("https", ignoreCase = true)) { "apiOrigin must use HTTPS" }
            require(origin.host != null && origin.rawUserInfo == null && origin.rawQuery == null && origin.rawFragment == null) {
                "apiOrigin must be an HTTPS origin or base path without credentials, query, or fragment"
            }
            val basePath = origin.rawPath.trimEnd('/')
            return "${origin.scheme}://${origin.rawAuthority}$basePath/api/v1/collect"
        }

        private fun SeeRayAnalyticsOptions.validated(): SeeRayAnalyticsOptions {
            require(siteId.matches(Regex("^[A-Za-z0-9][A-Za-z0-9._-]{0,63}$"))) { "siteId is invalid" }
            endpoint(apiOrigin)
            require(!flushInterval.isNegative && !flushInterval.isZero) { "flushInterval must be positive" }
            return copy(batchSize = batchSize.coerceIn(1, MAX_BATCH_SIZE))
        }

        private fun validatedPageUrl(value: String?): String? {
            if (value.isNullOrBlank() || value.length > 2048) return null
            return runCatching {
                URI(value).takeIf {
                    (it.scheme == "https" || it.scheme == "http") && it.host != null && it.rawUserInfo == null
                }?.let { uri ->
                    // Queries and fragments frequently contain search terms or account/reset tokens.
                    "${uri.scheme}://${uri.rawAuthority}${uri.rawPath?.ifEmpty { "/" } ?: "/"}"
                }
            }.getOrNull()
        }

        private fun cleanEventType(value: String): String? = value.trim().takeIf {
            it.length in 1..MAX_EVENT_TYPE && it.matches(Regex("^[A-Za-z0-9][A-Za-z0-9_.-]*$"))
        }

        private fun boundedText(value: String?, max: Int): String? = value
            ?.filterNot { it.isISOControl() }
            ?.trim()
            ?.take(max)
            ?.takeIf(String::isNotEmpty)

        private fun cleanUserId(value: String?): String? = boundedText(value, 128)

        private fun cleanProperties(input: Map<String, Any?>): Map<String, Any> = input.entries
            .asSequence()
            .take(10)
            .mapNotNull { (key, value) ->
                val cleanKey = boundedText(key, 32) ?: return@mapNotNull null
                val cleanValue: Any = when (value) {
                    is String -> boundedText(value, 128) ?: return@mapNotNull null
                    is Boolean -> value
                    is Byte, is Short, is Int, is Long -> (value as Number).toLong()
                    is Float, is Double -> (value as Number).toDouble().takeIf(Double::isFinite) ?: return@mapNotNull null
                    else -> return@mapNotNull null
                }
                cleanKey to cleanValue
            }
            .toMap(LinkedHashMap())

        private fun jsonObject(fields: Map<String, Any?>): String = fields.entries
            .filter { it.value != null }
            .joinToString(prefix = "{", postfix = "}") { (key, value) -> "${jsonString(key)}:${jsonValue(value)}" }

        private fun jsonValue(value: Any?): String = when (value) {
            null -> "null"
            is JsonFragment -> value.value
            is String -> jsonString(value)
            is Number -> value.toString()
            is Boolean -> value.toString()
            is Map<*, *> -> jsonObject(value.entries.associate { it.key.toString() to it.value })
            is Iterable<*> -> value.joinToString(prefix = "[", postfix = "]") { jsonValue(it) }
            else -> error("Unsupported JSON value: ${value.javaClass.name}")
        }

        private fun jsonString(value: String): String = buildString(value.length + 2) {
            append('"')
            value.forEach { character ->
                when (character) {
                    '"' -> append("\\\"")
                    '\\' -> append("\\\\")
                    '\b' -> append("\\b")
                    '\u000c' -> append("\\f")
                    '\n' -> append("\\n")
                    '\r' -> append("\\r")
                    '\t' -> append("\\t")
                    else -> if (character.code < 0x20) append("\\u%04x".format(character.code)) else append(character)
                }
            }
            append('"')
        }
    }
}

internal interface AnalyticsStore {
    fun get(key: String): String?
    fun put(key: String, value: String)
    fun remove(key: String)
}

internal interface AnalyticsTransport {
    fun post(endpoint: String, body: String): Boolean
}

private class AndroidAnalyticsStore(context: Context) : AnalyticsStore {
    private val preferences: SharedPreferences = context.getSharedPreferences("seeray_analytics", Context.MODE_PRIVATE)
    override fun get(key: String): String? = preferences.getString(key, null)
    override fun put(key: String, value: String) { preferences.edit().putString(key, value).apply() }
    override fun remove(key: String) { preferences.edit().remove(key).apply() }
}

private class UrlConnectionAnalyticsTransport : AnalyticsTransport {
    override fun post(endpoint: String, body: String): Boolean {
        val connection = URL(endpoint).openConnection() as HttpURLConnection
        return try {
            connection.requestMethod = "POST"
            connection.connectTimeout = 5_000
            connection.readTimeout = 5_000
            connection.doOutput = true
            connection.setRequestProperty("Content-Type", "application/json; charset=utf-8")
            connection.setFixedLengthStreamingMode(body.toByteArray(StandardCharsets.UTF_8).size)
            connection.outputStream.use { it.write(body.toByteArray(StandardCharsets.UTF_8)) }
            connection.responseCode in 200..299
        } finally {
            connection.disconnect()
        }
    }
}

File: sdk/ios/Sources/SeeRayAnalytics/SeeRayAnalytics.swift
import Foundation
#if canImport(FoundationNetworking)
    import FoundationNetworking
#endif
#if canImport(UIKit)
    import UIKit
#endif

protocol SeeRayAnalyticsStorage: Sendable {
    func string(forKey key: String) -> String?
    func set(_ value: String, forKey key: String)
    func removeValue(forKey key: String)
}

protocol SeeRayAnalyticsTransport: Sendable {
    func send(_ body: Data, to endpoint: URL) async throws
}

private final class UserDefaultsStorage: SeeRayAnalyticsStorage, @unchecked Sendable {
    private let defaults: UserDefaults
    private let lock = NSLock()

    init(defaults: UserDefaults) {
        self.defaults = defaults
    }

    func string(forKey key: String) -> String? {
        lock.lock()
        defer { lock.unlock() }
        return defaults.string(forKey: key)
    }

    func set(_ value: String, forKey key: String) {
        lock.lock()
        defer { lock.unlock() }
        defaults.set(value, forKey: key)
    }

    func removeValue(forKey key: String) {
        lock.lock()
        defer { lock.unlock() }
        defaults.removeObject(forKey: key)
    }
}

private final class URLSessionTransport: SeeRayAnalyticsTransport, @unchecked Sendable {
    private let session: URLSession

    init(session: URLSession) {
        self.session = session
    }

    func send(_ body: Data, to endpoint: URL) async throws {
        var request = URLRequest(url: endpoint)
        request.httpMethod = "POST"
        request.httpBody = body
        request.httpShouldHandleCookies = false
        request.cachePolicy = .reloadIgnoringLocalCacheData
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        request.setValue("no-store", forHTTPHeaderField: "Cache-Control")

        let (_, response) = try await session.data(for: request)
        guard let response = response as? HTTPURLResponse,
              (200..<300).contains(response.statusCode)
        else {
            throw URLError(.badServerResponse)
        }
    }
}

/// Explicit, consent-aware analytics for native iOS applications.
///
/// The SDK does not infer navigation, crashes, device models, advertising IDs, or user identity.
/// The host app decides when a screen becomes visible and when its conversion actually succeeds.
public actor SeeRayAnalytics {
    private static let maximumBatchSize = 10
    private static let maximumQueuedEvents = 100
    private static let sessionTimeout: TimeInterval = 30 * 60

    private let options: SeeRayAnalyticsOptions
    private let endpoint: URL
    private let storage: SeeRayAnalyticsStorage
    private let transport: SeeRayAnalyticsTransport
    private let now: @Sendable () -> Date
    private let batchSize: Int
    private let flushInterval: TimeInterval

    private var pending: [SeeRayTrackingEvent] = []
    private var flushTask: Task<Void, Never>?
    private var userId: String?
    private var closed = false
    private var droppedEventCount = 0

    public init(options: SeeRayAnalyticsOptions) throws {
        let configuration = URLSessionConfiguration.ephemeral
        configuration.httpCookieStorage = nil
        configuration.urlCredentialStorage = nil
        configuration.urlCache = nil
        try self.init(
            options: options,
            storage: UserDefaultsStorage(defaults: .standard),
            transport: URLSessionTransport(session: URLSession(configuration: configuration)),
            now: { Date() }
        )
    }

    init(
        options: SeeRayAnalyticsOptions,
        storage: SeeRayAnalyticsStorage,
        transport: SeeRayAnalyticsTransport,
        now: @escaping @Sendable () -> Date = { Date() }
    ) throws {
        guard options.siteId.range(
            of: #"^[A-Za-z0-9][A-Za-z0-9._-]{0,63}$"#,
            options: .regularExpression
        ) != nil else {
            throw SeeRayAnalyticsError.invalidSiteId
        }
        guard options.flushInterval.isFinite, options.flushInterval > 0 else {
            throw SeeRayAnalyticsError.invalidFlushInterval
        }
        guard let endpoint = Self.endpoint(for: options.apiOrigin) else {
            throw SeeRayAnalyticsError.invalidAPIOrigin
        }

        self.options = options
        self.endpoint = endpoint
        self.storage = storage
        self.transport = transport
        self.now = now
        self.batchSize = min(max(options.batchSize, 1), Self.maximumBatchSize)
        self.flushInterval = min(max(options.flushInterval, 0.2), 60)
    }

    public func consentState() -> SeeRayConsentState {
        if let saved = storage.string(forKey: consentKey),
           let state = SeeRayConsentState(rawValue: saved), state != .unknown
        {
            return state
        }
        return options.requireConsent ? .unknown : .granted
    }

    /// Apply the visitor's choice. Denial immediately clears queued events and SDK-owned IDs.
    public func setConsent(granted: Bool) {
        guard !closed else { return }
        storage.set(granted ? SeeRayConsentState.granted.rawValue : SeeRayConsentState.denied.rawValue,
                    forKey: consentKey)
        guard !granted else { return }

        pending.removeAll(keepingCapacity: false)
        flushTask?.cancel()
        flushTask = nil
        userId = nil
        storage.removeValue(forKey: visitorKey)
        storage.removeValue(forKey: sessionKey)
        storage.removeValue(forKey: lastActivityKey)
    }

    /// Store an opaque application account ID for subsequent consented events only.
    public func setUserId(_ value: String?) {
        guard !closed, consentState() == .granted else {
            userId = nil
            return
        }
        userId = Self.cleanText(value, maximumLength: 256)
    }

    public func trackPageView(url: String, title: String? = nil, referrer: String? = nil) {
        record(type: "page_view", url: url, title: title, referrer: referrer)
    }

    public func trackScreen(
        name: String,
        url: String,
        title: String? = nil,
        referrer: String? = nil
    ) {
        let screen = Self.cleanText(name, maximumLength: 120)
        guard let screen else { return }
        record(
            type: "page_view",
            url: url,
            title: title ?? screen,
            referrer: referrer,
            properties: ["screen": .string(screen)]
        )
    }

    public func trackEvent(
        type: String,
        url: String,
        category: String? = nil,
        action: String? = nil,
        name: String? = nil,
        properties: [String: SeeRayValue] = [:],
        title: String? = nil,
        referrer: String? = nil
    ) {
        record(
            type: type,
            url: url,
            title: title,
            referrer: referrer,
            category: category,
            action: action,
            name: name,
            properties: properties
        )
    }

    public func trackGoal(
        name: String,
        url: String,
        properties: [String: SeeRayValue] = [:],
        title: String? = nil
    ) {
        record(type: "goal", url: url, title: title, name: name, properties: properties)
    }

    /// Sends at most one bounded batch. Failed batches are requeued for a later flush.
    @discardableResult
    public func flush() async -> Bool {
        flushTask?.cancel()
        flushTask = nil
        guard consentState() == .granted else {
            pending.removeAll(keepingCapacity: false)
            return true
        }
        guard !pending.isEmpty else { return true }

        let count = min(batchSize, pending.count)
        let batchEvents = Array(pending.prefix(count))
        pending.removeFirst(count)
        let body = SeeRayTrackingBatch(
            schemaVersion: 1,
            siteId: options.siteId,
            sentAt: seeRayTimestamp(now()),
            events: batchEvents
        )

        do {
            let encoder = JSONEncoder()
            try await transport.send(encoder.encode(body), to: endpoint)
            storage.set(String(now().timeIntervalSince1970), forKey: lastActivityKey)
            if !pending.isEmpty, !closed { scheduleFlush() }
            return true
        } catch {
            guard consentState() == .granted else { return false }
            pending.insert(contentsOf: batchEvents, at: 0)
            if !closed { scheduleFlush() }
            return false
        }
    }

    /// Stop automatic scheduling and make one best-effort final send.
    public func close() async {
        guard !closed else { return }
        closed = true
        flushTask?.cancel()
        flushTask = nil
        _ = await flush()
    }

    public func queuedEventCount() -> Int { pending.count }
    public func droppedEventCountValue() -> Int { droppedEventCount }

    private func record(
        type: String,
        url: String,
        title: String?,
        referrer: String? = nil,
        category: String? = nil,
        action: String? = nil,
        name: String? = nil,
        properties: [String: SeeRayValue] = [:]
    ) {
        guard !closed, consentState() == .granted,
              let eventType = Self.cleanText(type, maximumLength: 64),
              let pageURL = Self.minimizedURL(url)
        else { return }
        guard pending.count < Self.maximumQueuedEvents else {
            droppedEventCount += 1
            return
        }

        let date = now()
        let identity = ensureIdentity(at: date)
        let cleanProperties = cleanProperties(properties)
        pending.append(
            SeeRayTrackingEvent(
                eventId: UUID().uuidString.lowercased(),
                type: eventType,
                occurredAt: seeRayTimestamp(date),
                url: pageURL,
                title: Self.cleanText(title, maximumLength: 256),
                referrer: referrer.flatMap(Self.minimizedURL),
                visitorId: identity.visitor,
                sessionId: identity.session,
                userId: userId,
                category: Self.cleanText(category, maximumLength: 120),
                action: Self.cleanText(action, maximumLength: 120),
                name: Self.cleanText(name, maximumLength: 120),
                properties: cleanProperties.isEmpty ? nil : cleanProperties,
                context: Self.context()
            )
        )
        scheduleFlush(immediate: pending.count >= batchSize)
    }

    private func ensureIdentity(at date: Date) -> (visitor: String, session: String) {
        let visitor = storage.string(forKey: visitorKey) ?? UUID().uuidString.lowercased()
        storage.set(visitor, forKey: visitorKey)

        let previousActivity = storage.string(forKey: lastActivityKey).flatMap(TimeInterval.init)
        let existingSession = storage.string(forKey: sessionKey)
        let elapsed = previousActivity.map { date.timeIntervalSince1970 - $0 }
        let rotate = existingSession == nil || elapsed == nil || (elapsed ?? 0) >= Self.sessionTimeout || (elapsed ?? 0) < 0
        let session = rotate ? UUID().uuidString.lowercased() : existingSession!
        storage.set(session, forKey: sessionKey)
        storage.set(String(date.timeIntervalSince1970), forKey: lastActivityKey)
        return (visitor, session)
    }

    private func scheduleFlush(immediate: Bool = false) {
        guard !closed, flushTask == nil else { return }
        let delay = immediate ? 0 : flushInterval
        let nanoseconds = UInt64(min(delay, 60) * 1_000_000_000)
        flushTask = Task { [weak self] in
            if nanoseconds > 0 {
                try? await Task.sleep(nanoseconds: nanoseconds)
            }
            guard !Task.isCancelled else { return }
            await self?.scheduledFlush()
        }
    }

    private func scheduledFlush() async {
        flushTask = nil
        _ = await flush()
    }

    private func cleanProperties(_ values: [String: SeeRayValue]) -> [String: SeeRayValue] {
        var result: [String: SeeRayValue] = [:]
        for (key, value) in values.prefix(30) {
            guard let cleanKey = Self.cleanText(key, maximumLength: 64),
                  let cleanValue = value.bounded
            else { continue }
            result[cleanKey] = cleanValue
        }
        return result
    }

    private static func cleanText(_ value: String?, maximumLength: Int) -> String? {
        guard let value else { return nil }
        let scalars = value.unicodeScalars.filter { !CharacterSet.controlCharacters.contains($0) }
        let cleaned = String(String.UnicodeScalarView(scalars)).trimmingCharacters(in: .whitespacesAndNewlines)
        guard !cleaned.isEmpty else { return nil }
        return String(cleaned.prefix(maximumLength))
    }

    private static func minimizedURL(_ value: String) -> String? {
        guard var components = URLComponents(string: value),
              let scheme = components.scheme?.lowercased(),
              scheme == "https" || scheme == "http",
              components.host != nil,
              components.user == nil,
              components.password == nil
        else { return nil }
        components.query = nil
        components.fragment = nil
        guard let minimized = components.string, minimized.count <= 2_048 else { return nil }
        return minimized
    }

    private static func context() -> SeeRayEventContext {
        let version = ProcessInfo.processInfo.operatingSystemVersion
        let osVersion = "\(version.majorVersion).\(version.minorVersion).\(version.patchVersion)"
        #if canImport(UIKit)
            let deviceType = UIDevice.current.userInterfaceIdiom == .pad ? "tablet" : "mobile"
        #else
            let deviceType = "mobile"
        #endif
        let language = Locale.preferredLanguages.first?.replacingOccurrences(of: "_", with: "-") ?? "und"
        return SeeRayEventContext(
            browser: "Other",
            operatingSystem: "iOS",
            operatingSystemVersion: String(osVersion.prefix(40)),
            deviceType: deviceType,
            language: String(language.prefix(40))
        )
    }

    private static func endpoint(for apiOrigin: String) -> URL? {
        guard var components = URLComponents(string: apiOrigin),
              components.scheme?.lowercased() == "https",
              components.host != nil,
              components.user == nil,
              components.password == nil,
              components.query == nil,
              components.fragment == nil
        else { return nil }
        let basePath = components.path.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        components.path = "/" + [basePath, "api/v1/collect"].filter { !$0.isEmpty }.joined(separator: "/")
        return components.url
    }

    private var consentKey: String { "seeray:\(options.siteId):consent" }
    private var visitorKey: String { "seeray:\(options.siteId):visitor_id" }
    private var sessionKey: String { "seeray:\(options.siteId):session_id" }
    private var lastActivityKey: String { "seeray:\(options.siteId):session_last_activity" }
}
